crt/fork.c

```diff
@@ -0,0 +1,569 @@
+{
+    struct pthread* self;
+    uintptr_t* dtv;
+    struct pthread* prev;
+    struct pthread* next;
+    uintptr_t sysinfo;
+    uintptr_t canary;
+    uintptr_t canary2;
+    int tid;
+    int errno_val;
+    volatile int detach_state;
+    volatile int cancel;
+    volatile unsigned char canceldisable;
+    volatile unsigned char cancelasync;
+    unsigned char tsd_used : 1;
+    unsigned char dlerror_flag : 1;
+    unsigned char* map_base;
+    size_t map_size;
+    void* stack;
+    size_t stack_size;
+    size_t guard_size;
+    void* result;
+    struct __ptcb* cancelbuf;
+    void** tsd;
+    struct
+    {
+        volatile void* volatile head;
+        long off;
+        volatile void* volatile pending;
+    } robust_list;
+    volatile int timer_id;
+    locale_t locale;
+    volatile int killlock[1];
+    char* dlerror_buf;
+    void* stdio_locks;
+    uintptr_t canary_at_end;
+    uintptr_t* dtv_copy;
+};
+
+int myst_round_up(uint64_t x, uint64_t m, uint64_t* r)
+{
+    uint64_t t;
+
+    if (!r)
+        return -EINVAL;
+
+    /* prevent divide by zero */
+    if (m == 0)
+        return -ERANGE;
+
+    if (__builtin_add_overflow(x, m - 1, &t))
+        return -ERANGE;
+
+    if (__builtin_mul_overflow(t / m, m, r))
+        return -ERANGE;
+
+    return 0;
+}
+
+/* ATTN: does not work for Linux yet: SYS_arch_prctl */
+static void _set_fsbase(void* p)
+{
+    __asm__ volatile("wrfsbase %0" ::"r"(p));
+}
+
+static inline struct pthread* __pthread_self()
+{
+    struct pthread* self;
+    __asm__("mov %%fs:0,%0" : "=r"(self));
+    return self;
+}
+
+extern hidden unsigned __default_guardsize;
+extern hidden unsigned __default_stacksize;
+extern hidden volatile size_t __pthread_tsd_size;
+
+extern hidden void* __copy_tls(unsigned char*);
+
+static void _get_parent_stack(void** stack, size_t* stack_size)
+{
+    struct pthread* self = __pthread_self();
+
+    if (self->stack && self->stack_size)
+    {
+        *stack = (uint8_t*)self->stack - self->stack_size;
+        *stack_size = self->stack_size;
+    }
+    else
+    {
+        const long n = SYS_myst_get_process_stack;
+
+        if (syscall(n, stack, stack_size) != 0)
+        {
+            fprintf(stderr, "cannot retrieve parent stack\n");
+            abort();
+        }
+    }
+}
+
+/*
+**==============================================================================
+**
+** _create_child_pthread()
+**
+** Create a new thread context from the current thread and return a pointer to
+** the new pthread structure. The pthread has the following layout.
+**
+**     [guard|stack|tls|tsd]
+**
+** Or:
+**     [ guard ]                td->guard_size | __default_guardsize
+**     [ stack ]                td->stack_size | __default_stacksize
+**     [ tls area | tcb ]       __libc.tls_size
+**     [ tsd area ]             __pthread_tsd_size
+**
+**==============================================================================
+*/
+
+struct pthread* _create_child_pthread()
```

Comment:
This function create child process stack and initialize the pthread structure. The function name is confusing.

Reply:
the stack is part of the pthread structure!

Reply:
or the pthread, stack, td. All in a single continuous thunk of memory